packages/ember-glimmer/tests/integration/syntax/each-in-test.js

```diff
@@ -57,222 +62,231 @@ applyMixins(BasicEachInTest,
+  createHash(/* hash */) {
+    throw new Error('Not implemented: `createHash`');
+  }
 
-    this.assertStableRerender();
+  makeHash(obj) {
+    let { hash, delegate } = this.createHash(obj);
 
-    this.runTask(() => {
-      set(this.context, 'categories.Smartphones', 100);
-      set(this.context, 'categories.Tweets', 443115);
-    });
+    this.hash = hash;
+    this.delegate = delegate;
+    return hash;
+  }
 
-    this.assertHTML(strip`
-      <ul>
-        <li>Smartphones: 100</li>
-        <li>JavaScript Frameworks: Infinity</li>
-        <li>Tweets: 443115</li>
-      </ul>
-    `);
+  replaceHash(hash) {
+    this.runTask(() => set(this.context, 'hash', this.createHash(hash).hash));
+  }
 
-    this.runTask(() => set(this.context, 'categories', {
-      'Smartphones': 8203,
-      'JavaScript Frameworks': Infinity
-    }));
+  clear() {
+    return this.runTask(() => set(this.context, 'hash', this.createHash({}).hash));
+  }
 
-    this.assertHTML(strip`
-      <ul>
-        <li>Smartphones: 8203</li>
-        <li>JavaScript Frameworks: Infinity</li>
-      </ul>
-    `);
+  setProp(key, value) {
+    return this.runTask(() => this.delegate.setProp(this.context, key, value));
   }
 
-  [`@test it can render sub-paths of each item`]() {
-    this.render(strip`
-      <ul>
-        {{#each-in categories as |category data|}}
-          <li>{{category}}: {{data.reports.unitsSold}}</li>
-        {{/each-in}}
-      </ul>
-    `, {
-      categories: {
-        'Smartphones': { reports: { unitsSold: 8203 } },
-        'JavaScript Frameworks': { reports: { unitsSold: Infinity } }
-      }
-    });
+  updateNestedValue(key, innerKey, value) {
+    return this.runTask(() => this.delegate.updateNestedValue(this.context, key, innerKey, value));
+  }
 
-    this.assertHTML(strip`
-      <ul>
-        <li>Smartphones: 8203</li>
-        <li>JavaScript Frameworks: Infinity</li>
-      </ul>
-    `);
+  render(template, context = {}) {
+    if (this.hash !== undefined) {
+      context.hash = this.hash;
+    }
+    if (this.type !== undefined) {
+      context.type = this.type;
+    }
+    context.secretKey = 'asd';
+
+    return super.render(template, context);
+  }
+}
+
+class EachInTest extends AbstractEachInTest {
+  [`@test it repeats the given block for each item in the hash`]() {
+    this.makeHash({ 'Smartphones': 8203, 'JavaScript Frameworks': Infinity });
+
+    this.render(`<ul>{{#each-in hash as |category count|}}<li>{{category}}: {{count}}</li>{{else}}Empty!{{/each-in}}</ul>`);
+
+    this.assertText('Smartphones: 8203JavaScript Frameworks: Infinity');
 
     this.assertStableRerender();
 
-    this.runTask(() => {
-      set(this.context, 'categories.Smartphones.reports.unitsSold', 100);
-      set(this.context, 'categories.Tweets', { reports: { unitsSold: 443115 } });
-    });
+    if (this.allowsSetProp) { // Not al backing data structures allow kvo tracking. Maps and Iterables don't
```

Comment:
@chancancode This is the idea I came up with to allow to run the same tests for every backing data structure but skip one specific part on data structures that do not allow KVO tracking: Maps and Iterables.

I considered (and I actually started with that approach) to define a `setProp` implementation that essentially replaced the Map/Iterable with a modified one, in an immutable fashino, but I had the impression that it was hiding the fact from the reader that the feature wasn't actually supported for some objects.